server/src/main/java/org/opensearch/index/store/RemoteSegmentStoreDirectory.java

```diff
@@ -0,0 +1,366 @@
+        }
+    }
+
+    /**
+     * Metadata of a segment that is uploaded to remote segment store.
+     */
+    static class UploadedSegmentMetadata {
+        private static final String SEPARATOR = "::";
+        private final String originalFilename;
+        private final String uploadedFilename;
+        private final String checksum;
+
+        UploadedSegmentMetadata(String originalFilename, String uploadedFilename, String checksum) {
+            this.originalFilename = originalFilename;
+            this.uploadedFilename = uploadedFilename;
+            this.checksum = checksum;
+        }
+
+        @Override
+        public String toString() {
+            return String.join(SEPARATOR, originalFilename, uploadedFilename, checksum);
+        }
+
+        public static UploadedSegmentMetadata fromString(String uploadedFilename) {
+            String[] values = uploadedFilename.split(SEPARATOR);
+            return new UploadedSegmentMetadata(values[0], values[1], values[2]);
+        }
+    }
+
+    /**
+     * Contains utility methods that provide various parts of metadata filename along with comparator
+     * Each metadata filename is of format: PREFIX__PrimaryTerm__Generation__UUID
+     */
+    static class MetadataFilenameUtils {
+        public static final String SEPARATOR = "__";
+        public static final String METADATA_PREFIX = "metadata";
+
+        /**
+         * Comparator to sort the metadata filenames. The order of sorting is: Primary Term, Generation, UUID
+         * Even though UUID sort does not provide any info on recency, it provides a consistent way to sort the filenames.
+         */
+        static class MetadataFilenameComparator implements Comparator<String> {
+            @Override
+            public int compare(String first, String second) {
+                String[] firstTokens = first.split(SEPARATOR);
+                String[] secondTokens = second.split(SEPARATOR);
+                if (!firstTokens[0].equals(secondTokens[0])) {
+                    return firstTokens[0].compareTo(secondTokens[0]);
+                }
+                long firstPrimaryTerm = getPrimaryTerm(firstTokens);
+                long secondPrimaryTerm = getPrimaryTerm(secondTokens);
+                if (firstPrimaryTerm != secondPrimaryTerm) {
+                    return firstPrimaryTerm > secondPrimaryTerm ? 1 : -1;
+                } else {
+                    long firstGeneration = getGeneration(firstTokens);
+                    long secondGeneration = getGeneration(secondTokens);
+                    if (firstGeneration != secondGeneration) {
+                        return firstGeneration > secondGeneration ? 1 : -1;
+                    } else {
+                        return getUuid(firstTokens).compareTo(getUuid(secondTokens));
+                    }
+                }
+            }
+        }
+
+        public static String getMetadataFilename(long primaryTerm, long generation, String uuid) {
```

Comment:
I would prefer this private as well.

Reply:
Makes sense. Maybe, need to keep package private for tests.